### ISSUE TITLE
Combined @auth rules (iam+userPools), accessing the items of others

### ISSUE THREAD

**Issue by mrnix** [aws-amplify/amplify-cli#2673]
**Describe the bug**
I have an issue with the combined authorization rules in app which allows to receive a feedback from Authenticated (userPools) and Unauthenticated (IAM) users (schema below). For some reason, Authenticated users can read/list only own-created items while Unauthenticated users can read/list all items in db (own + others).

**Amplify CLI Version**
3.17.0

**To Reproduce**
```graphql
type Feedback
  @model
  @auth(
    rules: [
      {allow: public, provider: iam, operations: [create, read, update, delete]}
      {allow: owner, operations: [create, read, update, delete]}
    ]
  ) {
  id: ID!
  identityId: String
  fileId: String
  resume: String
}
```

**Expected behavior**
Ideally I would like to set following access:
- Unauthenticated user can create and read their own item only (by accessing with an identityId taken by IAM)
- Authenticated users create one, can read all items

Do you have any idea how to make this schema better?

**Comment by mrnix** [comment]
In my case, Authenticated user (userPool) have incorrect result of `query listFeedback` where some random items from db are missed for some reason. I have 5 items in db, but get 3 items in result, despite the same rules. Super confusing and unpredictable.

Is there another place where custom rules are registered for each item in db?

**Comment by mrnix** [comment]
Another weird case: 
I have 2 items created by Authenticated owner, and 3 of Unauthenticated (iam) users.
Unauthenticated user can access to all items (it's ok)
Authenticated user can access to own items, but only see some 1 or 2 random item of other unauth authors.
But, once the Auth user create a new item, it can access to other's items that wasn't previously accessible.  It's an unobvious magic here

**Comment by hirochachacha** [comment]
I have the same issue and I think I can explain the reason.
If you open AppSync console and look in the schema, you'll see the reason.
@model generates multiple types implicitly. For example, 

```
type Post @model @aws_api_key {
}
```

generates

```
type Query {
  getPost(id: ID!): Post
  listPosts(filter: ModelPostFilterInput, limit: Int, nextToken: String): ModelPostConnection
}

type ModelPostConnection {
  items: [Post]
  nextToken: String
}

input ModelPostFilterInput {
}
```
for read operations.

However, it's not working in multi auth mode.

According to https://aws.amazon.com/blogs/mobile/using-multiple-authorization-types-with-aws-appsync-graphql-apis/, 

We actually need 
```
type Query {
  getPost(id: ID!): Post @aws_api_key
  listPosts(filter: ModelPostFilterInput, limit: Int, nextToken: String): ModelPostConnection @aws_api_key
}

type ModelPostConnection @aws_api_key {
  items: [Post]
  nextToken: String
}
```

instead.

In other words, current @model directive cannot handle multi auth directives well.

**Comment by hirochachacha** [comment]
Oh, never minds. I misunderstood that this is the same issue. I'll file the new issue instead.

**Comment by hirochachacha** [comment]
No, it's the same issue? You didn't use @aws_iam and @aws_cognito_user_pools directives?

**Comment by attilah** [comment]
@mrnix do you have a repro project what I could take a look at and provide complete mutations to repro those "random" query results, that seems off to me and I'd look in to it.

@hirochachacha if you're using ```@model``` then it goes best with ```@auth``` because the directives you used ```@aws*``` those are the service level directive and the CLI is just passing them through, the resolvers will have no code to handle those, this is by design.

**Comment by mrnix** [comment]
@attilah sure, 
here is schema: https://github.com/mrnix/importa/blob/2edb875c4f4c8fcd06caf7e184c8cd9bcfa11e2a/packages/backend/amplify/backend/api/importa/schema.graphql#L48

Query request, I finally make it work consistant for Auth and Unauth with passing the `authMode: 'AWS_IAM'` param. https://github.com/mrnix/importa/blob/2edb875c4f4c8fcd06caf7e184c8cd9bcfa11e2a/packages/web/src/components/ExportView/ExportView.tsx#L76

Thank you for your help!

**Comment by hirochachacha** [comment]
@attilah Thank you.  I can be making progress. However, I still see the problem for `@connection`.

```
type Post
  @model
  @auth(rules: [
    {provider: oidc, identityClaim: "sub",  allow: owner}
  ])
  @auth(rules: [
    {provider: apiKey, allow: public, operations: [read]}
  ]) {
    id: ID!
    text: String!
    createdAt: AWSDateTime

    comments: [Comment!] @connection(name: "PostComments", sortField: "createdAt")
}

type Comment
  @model
  @auth(rules: [
    {provider: oidc, identityClaim: "sub",  allow: owner}
  ])
  @auth(rules: [
    {provider: apiKey, allow: public, operations: [read]}
  ]) {
    id: ID!
    text: String!
    createdAt: AWSDateTime

    post: Post! @connection(name: "PostComments", sortField: "createdAt")
}
```

will generate

```
type Post @aws_api_key @aws_oidc {
  comments(createdAt: ModelStringKeyConditionInput, filter: ModelCommentFilterInput, limit: Int, nextToken: String, sortDirection:     ModelSortDirection): ModelCommentConnection @aws_api_key
  id: ID!
  text: String!
}
...
```

As you can see, `Post.comments` field has only `@aws_api_key`, doesn't have `@aws_oidc`, therefore `oidc` users cannot read `Post.comments` field. I guess this is a bug.

@mrnix 's approach is good work around for now.

**Comment by hirochachacha** [comment]
When I set `@aws_api_key` and `aws_oidc` explicitly, 

```
  ...
  comments: [Comment!] @connection(name: "PostComments", sortField: "createdAt") @aws_api_key @aws_oidc
  ...
```

I see the result contains `@aws_oidc`

```
  ...
  comments(createdAt: ModelStringKeyConditionInput, filter: ModelCommentFilterInput, limit: Int, nextToken: String, sortDirection:     ModelSortDirection): ModelCommentConnection @aws_api_key @aws_oidc
  ...
```

But it has no effect for access control. I have no idea.

**Comment by hirochachacha** [comment]
I was misunderstanding about @auth usage.

I end up with a usage like
```
  @model
  @auth(rules: [
    {allow: owner, provider: oidc, identityClaim: "sub"},
    {allow: groups, groups: ["user"], operations: [read], provider: oidc, groupClaim: "https://example.com/roles"}
  ])
  @auth(rules: [
    {allow: public, operations: [read], provider: apiKey}
  ])
```

It works fine except the fact that current implementation doesn't support groups authorization for oidc provider. Therefore, I've created https://github.com/aws-amplify/amplify-cli/pull/2719.
